CAN Message first pass

de846 opened the pull request:
This is a first pass at can messages, looking for feedback.

de846 committed:
First CAN Decoder implementation

de846 committed:
First CAN Decoder implementation

de846 committed:
First CAN Decoder implementation

de846 committed:
First CAN Decoder implementation

de846 commented on mercury/can.py:
```diff
@@ -0,0 +1,4 @@
+class CANDecoder(object):
```
This isn't being used yet, ignore for now. Will delete if not necessary.

de846 commented on mercury/can_decoder.py:
```diff
@@ -0,0 +1,46 @@
+    # we need the binary representation of the integer
+    msg = bin(int(msg))
+
+    msg = bin(int(example_can_msg, 2))
```
This override of `msg` is just here so I can test with the `example_can_msg` binary string above. Will be removed before merge.

de846 commented on mercury/can_sensors.py:
```diff
@@ -0,0 +1,17 @@
+SUSPENSION_BL = 11
+SUSPENSION_BR = 12
+
+CURRENT_FUEL_LEVEL = 13
```
I think we'll actually need a map the other direction, to have a number as the input (since it comes from the CAN ID field, and to map it to the proper sensor. I still am not sure how this will work with writing to the DB if these are separate, since for example, the model for wheel speed has all four sensors in a single row. We might need to allow nullable values and just write 4 separate rows for the "WheelSpeedSensor" unless we think of a better solution.

muneebafzal replied:
From my understanding, you assign a number for each sensor (like TEMP =1) and then in the CAN message you check if the data is of type 1 (TEMP =1) to know if the data package is about temperature? Please rectify my understanding

de846 replied:
What you describe (as I understand it now) is exactly what I have here, but I think I need the reversed mapping, since I'll know the number, and need to know the sensor it relates to. I think I have an idea of what to do.

de846 commented on mercury/test_can.py:
```diff
@@ -0,0 +1,10 @@
+from .can import CANDecoder
+
+
+class TestCANDecoder(TestCase):
```
I'll have to update this test to actually do something useful. Can be ignored for now.

de846 commented on mercury/views/can.py:
```diff
@@ -0,0 +1,18 @@
+        """Return 400 Bad Request if no CAN message is received"""
+        return HttpResponse(status=400)
+    log.debug(request.body)
+    log.debug(dir(request.body))
```
These will be removed, just still testing everything.

muneebafzal replied:
fair

de846 commented on mercury/views/can.py:
```diff
@@ -0,0 +1,18 @@
+    log.debug(request.body)
+    log.debug(dir(request.body))
+    decoded_message = decode_can_message(request.body)
+    return HttpResponse(decoded_message, status=201)
```
For right now, I am just running the input through the `decode_can_message` function and returning it back to the user with a 201. It is not actually writing anything yet. My intention is to figure out which sensor to write a new data row of based on the ID in the can message, so that will happen in this `post` method.

muneebafzal commented on mercury/can_decoder.py:
```diff
@@ -0,0 +1,46 @@
+    # the control field is a 6-bit field that contains the length of the
+    # data in bytes, so read n bits where n is 8 * control_field
+    control_field, msg = read_bits(msg, 6)
+    data, msg = read_bits(msg, int(control_field, 2) * 8)
```
I am not sure how and where you compare the given values of different sensors (from can_sensors.py, e.g TEMPERATURE=1) in decoding the CAN message? can you please help me understand?

de846 replied:
I probably need to re-work the can_sensors a bit, but the idea is that if we map the numeric ID to the sensor, we have the numeric value in the `msg_id` field on line 25.

de846 committed:
Update CAN decoder

de846 committed:
Remove unnecessary can_sensors.py

de846 committed:
Remove unnecessary can_decoder.py (in can.py now)

Features have been moved to can.py

de846 committed:
Moved to tests/test_can.py

de846 committed:
Add comments to rationale of CAN Decoder

Add extended ID support

de846 committed:
Add a Testing CAN Decoder that returns all fields
Add exception handling for constraints
Updated CAN Tests

de846 committed:
Add better CAN tests and return full dict method

de846 committed:
Add a form for the CAN Decoder

de846 committed:
use DEBUG instead of ERROR

de846 committed:
Update test for the new return type

de846 commented:
There's still some updates I'd like to do to clean this up, but I think we could demo it as-is. Gettingit merged will also mean that someone might be able to make it look pretty by demo time.

de846 committed:
Fix can decoder to handle multiple input types

Add tests to ensure correctness

de846 committed:
Check the string length so it isn't above 130
Don't return a server error (500) but instead return a bad request (400)
message to the user.

de846 committed:
Raise NotImplemented if no matching CAN ID

de846 committed:
Move <pre> tag lower and add result id and class

MichaelLally reviewed (approved):
LGTM
